Accusations manager

ghost opened the pull request:
## Scope

Migrating over from old repository.

Closes #421 
Closes #383 
Closes #373 
Closes #364

ricardopinto committed:
Starting structure of accusations manager

ricardopinto committed:
Couple edits and improvements

ricardopinto committed:
Polling from DB implementation

ricardopinto committed:
Integrated accusation manager with monitor

ricardopinto committed:
Polling is working, but it ran to a block that had no round states

ricardopinto committed:
Polling working with last 2 epochs of data

ricardopinto committed:
Accusation manager working with workers

ricardopinto committed:
Cache cleanup on accusation manager.
Improved change management.
Initial work on DB persistance.

ricardopinto committed:
Persistence is working with "encoding/gob" lib

ricardopinto committed:
Accusation manager finished

ricardopinto committed:
Fixed imports and ran npm audit fix to address critical security issues

ricardopinto committed:
Added more docs and improved tests

ricardopinto committed:
Added more docs and fixed a test

ricardopinto committed:
progress in accusation of multiple proposal

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto committed:
Merge branch 'accusations-manager' into double-proposal-accusation

ricardopinto committed:
Implemented multiple proposal accusation detection

ricardopinto committed:
* Improved multiple proposal detection
* Added placeholder for deterministic accusation ID
* Started unit testing of multiple proposal accusation

madhivedan commented on consensus/objs/accusations_test.go:
```diff
@@ -0,0 +1 @@
+package objs
```
Empty file should be removed.

ricardopinto replied:
Done.

madhivedan commented on consensus/synchronizer.go:
```diff
@@ -518,4 +521,21 @@ func (s *Synchronizer) setupLoops() {
+
+	accusationManagerLoopConfig := newLoopConfig().
+		withName("AccusationManagerLoop").
+		//withInitialDelay(9*constants.MsgTimeout).
```
Remove commented out code.

ricardopinto replied:
✅

madhivedan commented on consensus/synchronizer.go:
```diff
@@ -217,7 +219,7 @@ type Synchronizer struct {
 
 // Init initializes the struct
-func (s *Synchronizer) Init(cdb *db.Database, mdb *badger.DB, tdb *badger.DB, gc *gossip.Client, gh *gossip.Handlers, ep *evidence.Pool, eng *lstate.Engine, app *application.Application, ah *admin.Handlers, pman *peering.PeerManager, storage dynamics.StorageGetter) {
+func (s *Synchronizer) Init(cdb *db.Database, mdb *badger.DB, tdb *badger.DB, gc *gossip.Client, gh *gossip.Handlers, ep *evidence.Pool, eng *lstate.Engine, app *application.Application, ah *admin.Handlers, pman *peering.PeerManager, aman *accusation.Manager, storage dynamics.StorageGetter) {
```
At this point with this many fields (especially with multiple of the same type so you can't just rely on type checking) it probably makes sense to create a configuration struct to pass into this function.

ricardopinto replied:
Interesting idea for a future refactoring.

madhivedan commented on consensus/objs/accusations.go:
```diff
@@ -0,0 +1,68 @@
+
+const (
+	// Created means the accusation has been identified but not yet persisted in any way.
+	Created AccusationState = 0
```
Might want to use [iota](https://github.com/golang/go/wiki/Iota) for enum-like consts. Also, not a bad idea to have an `Unset/Invalid AccusationState` for the 0 value otherwise uninitialized values will be considered `Created`

ricardopinto replied:
✅

madhivedan commented on consensus/objs/accusations.go:
```diff
@@ -0,0 +1,68 @@
+	Completed AccusationState = 3
+)
+
+type Accusation interface {
```
Is it expected that there will be multiple implementers for an `Accusation`? If not, we can probably skip the interface, which would allow the removal of the getters and setters below.

diligentis replied:
Makes sense, thanks!

madhivedan commented on consensus/objs/accusationkey.go:
```diff
@@ -0,0 +1,68 @@
+// MarshalBinary takes the AccusationKey object and returns
+// the canonical byte slice
+func (a *AccusationKey) MarshalBinary() ([]byte, error) {
+	if a == nil {
```
`nil` pointer calls should be allowed to panic. It'll allow to simplify the return to just `[]byte`

ricardopinto replied:
Looking at other marshalling code in different structs we can see this pattern all over the place.

ricardopinto replied:
panic is the way to go 👍 

✅

madhivedan commented on consensus/accusation/manager.go:
```diff
@@ -0,0 +1,367 @@
+	closeChan := make(chan struct{}, 1)
+
+	m := &Manager{}
+	err := m.Init(database, sstore, logger, detectorLogics, workQ, accusationQ, closeChan)
```
Should just bring `Init` into `NewManager`, as there's no reason to have it as a separate exported method. Also it never returns an error, so this is extraneous error checking.

ricardopinto replied:
✅

madhivedan reviewed (commented):
Some nits, overall good work though!

ricardopinto committed:
Tested multiple proposal detection

ricardopinto committed:
Implemented deterministic accusation ID and started testing

vtleonardo committed:
Start adding subscription to the scheduler

kosegor committed:
change structs and naming

kosegor committed:
task schedule and kill refactor

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto committed:
Reviewed comments from PR

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto commented on consensus/accusation/detect_double_spend.go:
```diff
@@ -0,0 +1,11 @@
+
+func detectDoubleSpend(rs *objs.RoundState) (objs.Accusation, bool) {
+
+	return nil, false
```
Removed the files and the tickets already exist.

diligentis replied:
Thanks!

ricardopinto commented on consensus/accusation/detect_multiple_proposal.go:
```diff
@@ -0,0 +1,18 @@
+	// proposer []byte
+	// make sure proposer is a validator
+	// submit both proposals and already validated that both RClaims are valid and sigs are different
+	return nil, false
```
Removed the files and the tickets already exist.

diligentis replied:
Thanks!

ricardopinto committed:
improving accusation IDs

ricardopinto committed:
Improved IDs from UUID type to [32]byte

ricardopinto committed:
making sure UUIDs are still fine as ID

ricardopinto committed:
multiple proposal detection migration

ricardopinto committed:
Added RClaims.Equal

ricardopinto committed:
Added IsProposerValid() to lstate.RoundStates

ricardopinto committed:
Migrated old branch changes

kosegor committed:
fix internal communication with the task

ricardopinto committed:
Implemented multiple proposal accusation task which calls the smart contract

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto committed:
Changed error handling to panic on marshalling nil structs

ricardopinto committed:
Updated accusation to slash the dishonest validator

ricardopinto committed:
- Improved test coverage
- Fixed a bug in accusation key unmarshalling
- removed unused method db.DeleteAccusation()

kosegor committed:
big renaming refactor and executor response added to recovery

kosegor committed:
refactor validator, monitor and events in order to use the new tasks Handler

kosegor committed:
adding responses cache

kosegor committed:
json annotation

ricardopinto committed:
CHanged marshallers to avoid cyclic dependencies

ricardopinto committed:
Merge branch 'pr/vtleonardo/67' into accusations-manager-task-integration

ricardopinto committed:
Revert "Merge branch 'pr/vtleonardo/67' into accusations-manager-task-integration"

This reverts commit 9d64c42c6dd936b5314b8b8864ffb9119d73bd5e, reversing
changes made to 8c1b1573eaa4f9a0983f0128c70e227f4a7fa2c6.

kosegor committed:
merge from main

kosegor committed:
fix responses bug

kosegor committed:
fix task_executor tests and circular dependency

kosegor committed:
starting with handler and manager tests

ricardopinto committed:
Revert "Revert "Merge branch 'pr/vtleonardo/67' into accusations-manager-task-integration""

This reverts commit d5f4b93593be281adbfe3e4f7fd3f23665f1c8a2.

ricardopinto committed:
Merge branch 'add-subscription-scheduler' into accusations-manager-task-integration

ricardopinto committed:
Fixed compilation errors

kosegor committed:
more tests, recovery isnt fixed yet

ricardopinto committed:
Integration almost finished. Now refactoring tests

ricardopinto committed:
Tests passing

kosegor committed:
recovery fixed

ricardopinto committed:
Merge branch 'accusations-manager-task-integration' into multiple-proposal-detection

ricardopinto committed:
Updated multiple proposal detection to conform with integrations against new task scheduler

kosegor committed:
race condition fixed

kosegor committed:
merge from main

kosegor committed:
fix ethdkg tests

kosegor committed:
lint fixes

ricardopinto committed:
Added debug info to detection and attached multiple proposal detection to the pipeline.

ricardopinto committed:
Merge branch 'add-subscription-scheduler' into accusations-manager-task-integration

ricardopinto committed:
Merge branch 'accusations-manager-task-integration' into multiple-proposal-detection

kosegor committed:
moving clone to TM

ricardopinto committed:
make generate changes

ricardopinto committed:
Testing accusations

ricardopinto committed:
Accusations are already submitted to smart contracts

vtleonardo committed:
Merge remote-tracking branch 'upstream/main' into add-subscription-scheduler

kosegor committed:
fix bug with mutex

kosegor committed:
typo

kosegor committed:
addressing some of PR comments

kosegor committed:
fixed close for TM and added some tets

ricardopinto committed:
Fixed RBAC with accusation smart contracts

ricardopinto committed:
Merge branch 'add-subscription-scheduler' into multiple-proposal-detection

kosegor committed:
addressing PR comments

kosegor committed:
merge from main

Commit:
Fixing some weirdness with golangci-linter in CI

ricardopinto committed:
Accusing with smart contracts works

kosegor committed:
parallel some tests

kosegor committed:
merge from main

kosegor committed:
fix scripts bug

kosegor committed:
merge from main

kosegor committed:
fix version_check tests

ricardopinto committed:
Improved deterministic accusation ID

ricardopinto committed:
Merge branch 'add-subscription-scheduler' into accusations-manager-task-integration

ricardopinto committed:
Updated generated files

ricardopinto committed:
Merge branch 'accusations-manager-task-integration' into multiple-proposal-detection

ricardopinto committed:
Generated bindings and fixed code after breaking changes

ricardopinto committed:
Updated hardhat scripts to accomodate salt deployment of accusation contracts

ricardopinto committed:
fixed tests

kosegor committed:
merge from main

ricardopinto committed:
picked changes from child branch

kosegor committed:
Merge remote-tracking branch 'upstream/main' into add-subscription-scheduler

ricardopinto committed:
updated bindings

ricardopinto committed:
Merge branch 'accusations-manager' into my-accusations-manager

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto committed:
Clean up

ricardopinto committed:
Merge branch 'add-subscription-scheduler' into accusations-manager

ricardopinto committed:
It is not compiling

ricardopinto committed:
Merge branch 'accusations-manager' into multiple-proposal-detection

ricardopinto committed:
it now compiles again

ricardopinto committed:
Improved concurrency, caching and locking when processing local Round States

ricardopinto committed:
Improved concurrency, caching and locking when processing local Round States (#325)

ricardopinto committed:
Merge branch 'accusations-manager' into multiple-proposal-detection

diligentis commented:
Per convo with @Cr0wn-Gh0ul and @vtleonardo marking this as draft until after merge conflicts resolved.

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto committed:
updated bindings

ricardopinto committed:
Merge branch 'accusations-manager' into origin-accusations-manager

ricardopinto committed:
changed default accusations logger from debug to info

ricardopinto committed:
bridge tests are now passing

ricardopinto committed:
fixed solidity lint

ricardopinto committed:
fixed more linting

ricardopinto committed:
Merge branch 'origin-accusations-manager' into multiple-proposal-detection

ricardopinto committed:
Merge branch 'main' into origin-accusations-manager

ricardopinto committed:
Fixed DB prefix conflict

ricardopinto committed:
Merge branch 'origin-accusations-manager' into multiple-proposal-detection

ricardopinto committed:
Trying to ignore evicted validator in consensus

vtleonardo committed:
Add logic to handle ethdkg restart after slashing

kosegor committed:
restart ethdkg after eviction and snapshot

kosegor committed:
eviction with min validators fixed

kosegor committed:
fixes

kosegor committed:
ready to merge

kosegor committed:
Merge pull request #2 from ricardopinto/validator-eviction

Validator eviction

ricardopinto committed:
Fixed gitignore

ricardopinto committed:
Merge branch 'main' into accusations-manager

ricardopinto committed:
Added missing function getBytes32Salt

ricardopinto committed:
It now compiles

ricardopinto committed:
fixed tests

ricardopinto committed:
fixed typescript errors

ricardopinto committed:
Fixed deployment

ricardopinto committed:
linting fixed

ricardopinto committed:
fixed test

ricardopinto committed:
Cleanups and documentation

ricardopinto committed:
more test coverage

kosegor committed:
coverage for accusation manager

kosegor committed:
refresh

kosegor committed:
tests fixes
